tests/test_endpoint.cc

```diff
@@ -0,0 +1,137 @@
+ * wrappers requires a lot of extra support code.
+ *
+ * See test_endpoint.py for remaining unit tests.
+ */
+namespace lsst {
+namespace afw {
+namespace geom {
+
+namespace {
+/**
+ * An IcrsCoordEndpoint whose sole purpose is to be a nontrivial subclass
+ * of IcrsCoordEndpoint.
+ */
+class GratuitousIcrsCoordEndpoint : public IcrsCoordEndpoint {
+    // Default constructor is adequate
+};
+
+/**
+ * An IcrsCoordEndpoint that adds equality-relevant fields to IcrsCoordEndpoint.
+ */
+class CoolIcrsCoordEndpoint : public IcrsCoordEndpoint {
+public:
+    CoolIcrsCoordEndpoint(string const & coolness) : IcrsCoordEndpoint(), _coolness(coolness) {}
+    string getCoolness() const noexcept { return _coolness; }
+
+    bool operator==(BaseEndpoint const & other) const noexcept override {
+        if (!IcrsCoordEndpoint::operator==(other)) {
+            return false;
+        } else {
+            // Guaranteed not to throw by BaseEndpoint::operator==
+            auto coolOther = dynamic_cast<CoolIcrsCoordEndpoint const &>(other);
+            return _coolness == coolOther._coolness;
+        }
+    }
+
+private:
+    string const _coolness;
+};
+}
+
+/// Test whether Endpoint equality treats subclasses as non-substitutable.
+BOOST_AUTO_TEST_CASE(EndpointEqualsNotPolymorphic) {
+    IcrsCoordEndpoint superclass;
+    GratuitousIcrsCoordEndpoint subclass;
+
+    // Test both IcrsCoordEndpoint::operator== and GratuitousIcrsCoordEndpoint::operator==
+    BOOST_TEST(superclass != subclass);
+    BOOST_TEST(subclass != superclass);
+}
+
+string printEquals(bool equal, string const & name1, string const & name2) {
+    return name1 + (equal ? " == " : " != ") + name2;
+}
+
+/**
+ * Test whether Endpoint equality is symmetric and transitive even in the
+ * presence of subclasses that add value fields.
+ */
+BOOST_AUTO_TEST_CASE(EndpointEqualsAlgebraic) {
+    using namespace std::string_literals;
+
+    vector<std::shared_ptr<IcrsCoordEndpoint const>> endpoints = {
+        make_shared<CoolIcrsCoordEndpoint>("Very cool"s),
+        make_shared<CoolIcrsCoordEndpoint>("Slightly nifty"s),
+        make_shared<CoolIcrsCoordEndpoint>("Very cool"s),
+        make_shared<IcrsCoordEndpoint>()
+    };
+
+    BOOST_TEST(*(endpoints[0]) != *(endpoints[1]));
+    BOOST_TEST(*(endpoints[1]) != *(endpoints[0]));
+    BOOST_TEST(*(endpoints[0]) == *(endpoints[2]));
+    BOOST_TEST(*(endpoints[2]) == *(endpoints[0]));
+
+    for (auto const endpoint1 : endpoints) {
+        for (auto const endpoint2 : endpoints) {
+            for (auto const endpoint3 : endpoints) {
+                bool const equals12 = *endpoint1 == *endpoint2;
+                bool const equals13 = *endpoint1 == *endpoint3;
+                bool const equals23 = *endpoint2 == *endpoint3;
+
+                // Iteration covers all permutations, so no need to be redundant
+                bool const transitive = !equals12 || !equals13 || equals23;
```

Comment:
Can you please try to make the test easier to follow?  For a given combination, it isn't sufficient. For instance 1!=2 is enough to make it true, so even if 1==3 and 2==3 this will pass. I suspect it works by the time you iterate over all possible combinations, but it's not obvious to me that is so.

Reply:
I'm not sure I can. Transitivity is a fundamentally difficult assertion to test because it's an implication: the only way to verify it is by trying lots of combinations of objects (arguably I don't have enough) and seeing that the equalities and inequalities always line up.

I can add the missing permutations, but will that make the test easier or harder to follow?

Reply:
Perhaps some comments might help. I am not even sure what this particular line is testing.

Reply:
// Test endpoint1 == endpoint2 => endpoint2 == endpoint3
                // Iteration covers all permutations, so no need to be redundant
                bool const transitive = !equals12 || !equals13 || equals23;